extensions/control-plane/api/management-api/contract-agreement-api/src/main/java/org/eclipse/edc/connector/api/management/contractagreement/ContractAgreementNewApiController.java

```diff
@@ -0,0 +1,97 @@
+import org.eclipse.edc.connector.spi.contractagreement.ContractAgreementService;
+import org.eclipse.edc.jsonld.spi.JsonLd;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.result.Result;
+import org.eclipse.edc.transform.spi.TypeTransformerRegistry;
+import org.eclipse.edc.web.spi.exception.InvalidRequestException;
+import org.eclipse.edc.web.spi.exception.ObjectNotFoundException;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.edc.web.spi.exception.ServiceResultHandler.exceptionMapper;
+
+@Produces({ MediaType.APPLICATION_JSON })
+@Path("/v2/contractagreements")
+public class ContractAgreementNewApiController implements ContractAgreementNewApi {
+    private final ContractAgreementService service;
+    private final JsonLd jsonLdService;
+    private final TypeTransformerRegistry transformerRegistry;
+    private final Monitor monitor;
+
+    public ContractAgreementNewApiController(ContractAgreementService service, JsonLd jsonLdService, TypeTransformerRegistry transformerRegistry, Monitor monitor) {
+        this.service = service;
+        this.jsonLdService = jsonLdService;
+        this.transformerRegistry = transformerRegistry;
+        this.monitor = monitor;
+    }
+
+
+    @POST
+    @Path("/request")
+    @Override
+    public List<JsonObject> queryAllAgreements(@Valid QuerySpecDto querySpecDto) {
+        querySpecDto = ofNullable(querySpecDto).orElseGet(() -> QuerySpecDto.Builder.newInstance().build());
+
+        var query = transformerRegistry.transform(querySpecDto, QuerySpec.class)
+                .orElseThrow(InvalidRequestException::new);
+
+
+        try (var stream = service.query(query).orElseThrow(exceptionMapper(ContractDefinition.class, null))) {
+            return stream
+                    .map(it -> transformerRegistry.transform(it, ContractAgreementDto.class)
+                            .compose(dto -> transformerRegistry.transform(dto, JsonObject.class))
+                            .compose(jsonLdService::compact))
+                    .peek(r -> r.onFailure(f -> monitor.warning(f.getFailureDetail())))
+                    .filter(Result::succeeded)
+                    .map(Result::getContent)
+                    .collect(toList());
+        }
+    }
+
+    @GET
+    @Path("{id}")
+    @Override
+    public JsonObject getAgreementById(@PathParam("id") String id) {
+        return Optional.of(id)
+                .map(service::findById)
+                .map(it -> transformerRegistry.transform(it, ContractAgreementDto.class)
+                        .compose(dto -> transformerRegistry.transform(dto, JsonObject.class))
+                        .compose(jsonLdService::compact)
+                        .orElseThrow(InvalidRequestException::new))
```

Comment:
In this case it's not a request issue, but likely this should be mapped as an `EdcException`, that will result in a 500 error